numpy/core/tests/test_arrayprint.py

```diff
@@ -796,6 +797,49 @@ def test_dtype_linewidth_wrapping(self):
             array(['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1'],
                   dtype='{}')""".format(styp)))
 
+    @pytest.mark.parametrize(
+        ['native'],
+        [
+            ('bool',),
+            ('uint8',),
+            ('uint16',),
+            ('uint32',),
+            ('uint64',),
+            ('int8',),
+            ('int16',),
+            ('int32',),
+            ('int64',),
+            ('float16',),
+            ('float32',),
+            ('float64',),
+            ('<U1',),     # 4-byte width string
+        ],
+    )
+    def test_dtype_endianness_repr(self, native):
+        '''
+        there was an issue where 
+        repr(array([0], dtype='<u2')) and repr(array([0], dtype='>u2'))
+        both returned the same thing:
+        array([0], dtype=uint16)
+        even though their dtypes have different endianness.
+        '''
+        if native == '<U1' and sys.byteorder == 'big':
+            native = '>U1'
+        native_dtype = np.dtype(native)
+        non_native_dtype = native_dtype.newbyteorder()
+        non_native_repr = repr(np.array([1], non_native_dtype))
+        native_repr = repr(np.array([1], native_dtype))
+        # preserve the sensible default of only showing dtype if nonstandard
+        assert ('dtype' in native_repr) ^ (native_dtype in _typelessdata),\
+                ("an array's repr should show dtype if and only if the type "
+                 'of the array is NOT one of the standard types '
+                 '(e.g., int32, bool, float64).')
```

Comment:
I would prefer this as an `if native_dtype in [np.int_, np.float64, np.complex128, np.bool_]` (or `_tyeplessdata`), but that is just me nitpicking (xor is much harder to reason about).

Reply:
If you'd prefer something like
```py
msg = ("an array's repr should show dtype if and only if the type "
                 'of the array is NOT one of the standard types '
                 '(e.g., int32, bool, float64).')
if native_dtype in _typelessdata:
    assert 'dtype' in native_repr, msg
else:
    assert 'dtype' not in native_repr, msg
```
I see nothing wrong with that. That doesn't seem much more clear to me, but maybe I'm the only one.

Reply:
To me it is a bit nicer, especially if the `msg` was a comment on the branch.  But we already cleaned up the test a lot and this doesn't matter enough.

Thanks @molsonkiko it is a nice improvement/fix.